File: algorithms/hard/my_calendar_iii.ex
defmodule MyCalendarThree do
  use GenServer

  @spec init_() :: any
  def init_() do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
    GenServer.call(__MODULE__, :init)
  end

  @spec book(start :: integer, end_ :: integer) :: integer
  def book(start, end_) do
    GenServer.call(__MODULE__, {:book, start, end_})
  end

  def init(state) do
    {:ok, state}
  end

  def handle_call(:init, _from, _state) do
    {:reply, true, :orddict.new()}
  end

  def handle_call({:book, start, end_}, _from, state) do
    state = :orddict.update(start, &(&1 + 1), 1, state)
    state = :orddict.update(end_, &(&1 - 1), -1, state)

    state
    |> :orddict.to_list()
    |> Enum.reduce({0, 0}, fn {_, v}, {k, sum} ->
      {max(k, sum + v), sum + v}
    end)
    |> then(fn {k, _} -> {:reply, k, state} end)
  end
end

File: algorithms/hard/my_calendar_iii.rs
use std::collections::BTreeMap;

struct MyCalendarThree {
    map: BTreeMap<i32, i32>,
}

impl MyCalendarThree {
    fn new() -> Self {
        Self {
            map: BTreeMap::new(),
        }
    }

    fn book(&mut self, start: i32, end: i32) -> i32 {
        *self.map.entry(start).or_insert(0) += 1;
        *self.map.entry(end).or_insert(0) -= 1;
        let mut k = 0;
        let mut sum = 0;
        for v in self.map.values() {
            sum += *v;
            k = k.max(sum);
        }
        k
    }
}
